cpp/include/cudf/column/column_view.hpp

```diff
@@ -331,7 +341,8 @@ class column_view : public detail::column_view_base {
  private:
   std::vector<column_view> _children{};  ///< Based on element type, children
                                          ///< may contain additional data
-};                                       // namespace cudf
+  std::shared_ptr<column_view> _dictionary_keys;    ///< Keys for dictionary column
```

Comment:
So a `column_view` variable cannot be a member of a `column_view` class. It needs to be a pointer of contained inside of something else. So it needs to live somewhere (at least the first one does).

Reply:
Oh yeah. In that case, it could just be a raw `column_view*`.

Reply:
> Oh yeah. In that case, it could just be a raw `column_view*`.

But where would the instance live? That is, who would own the life-time of this object?
I believe a raw pointer would require coding the `copy-ctors` and `operator=` functions while a `shared_ptr` would work with the default implementation.

Reply:
Oh bother, you're right. Hm, keep the `shared_ptr` for now and I'll chew on it a bit.

Reply:
On another note: Why is `dictionary_keys` stored separately from child columns? Does it not make sense that a dictionary column's first child would be the keys?

Reply:
> The column_view cannot work without it's column

You can have a `column_view` without a `column`. 

I think the confusion about that point explains the confusion here:

> I have been confused why we need a column_view. 

It may be helpful to review `std::string_view` and why it exists instead of just passing `std::string const&` everywhere: https://abseil.io/tips/1

Reply:
I'm not sure the use case for `string_view` lines up exactly with the `column_view` use case, but I can see how having an `offset` for `string_view` makes things more efficient - and the alternatives cause even more overhead.

That being said, I haven't had any particular reason to dislike `column_view` until I saw the use of `shared_ptr<column_view>`. It sounds like the reason for a shared pointer in this case is that 1) `column_view` can't have a recursively defined member `column_view`, and `column_view*` fails to denote ownership well.

Are the dictionary keys owned by this column? If so, we can avoid the use of `shared_ptr` by putting the dictionary keys in the children vector.

Reply:
> I'm not sure the use case for string_view lines up exactly with the column_view use case

It's very nearly the same thing. 

`string_view` allows you to zero-copy operate on either a `std::string` or a `const char*`.

`column_view` allows you to zero-copy operate on either a `cudf::column` or other raw, device memory. 

The point is that `column_view` enables operating on device memory allocated outside of libcudf. If all we had is `column`, that would force people to deep copy any memory they already allocated into a `column` in order for `libcudf` to work on it.

Reply:
@jrhemstad I think we're on the same page. I'm just thinking out loud - there is more flexibility in how we can implement views over columns than there is on how the stl can implement views over char*, but they are accomplishing the same thing.